### PR TITLE
Added functions for dimension and maximal independent set over fields

### PR DESCRIPTION
**Pull request by raulepure**
Both work over Singular Fields. Tests and Documentation are also added.
Also added "maGetPerm" on C-level to be able to create Singular internal maps between
polynomial rings over fields and polynomial rings over function fields.
Changed "p_PermPoly" to be able to work with function fields.

### CODE REVIEW

File: deps/src/ideals.cpp [hunk 1]
```diff
@@ -240,4 +240,35 @@ void singular_define_ideals(jlcxx::Module & Singular)
...
+        lists L = scIndIndset(I, all, r->qideal);
+        int n = rVar(r);
+        if(all == true)
+        {
+           int m = lSize(L);
+           for(int i = 0; i<m; i++)
+           {
+              intvec * v = reinterpret_cast<intvec *>(L->m[i].data);
+              int * content = v->ivGetVec();
+              for(int j=0; j<n; j++)
```
Comment (wbhart) [inline deps/src/ideals.cpp]: space before and after < and =

File: deps/src/ideals.cpp [hunk 2]
```diff
@@ -240,4 +240,35 @@ void singular_define_ideals(jlcxx::Module & Singular)
...
+              {
+                 a.push_back(content[j]);
+              }
+           }
+        }
+        else
+        {
+           intvec * v = reinterpret_cast<intvec *>(L->m[0].data);
+           int * content = v->ivGetVec();
+           for(int j=0; j<n; j++)
```
Comment (wbhart) [inline deps/src/ideals.cpp]: spacing

File: deps/src/rings.cpp [hunk 3]
```diff
@@ -282,10 +282,23 @@ void singular_define_rings(jlcxx::Module & Singular)
...
-                                     ring new_ring, void * map_func_ptr) {
+                                     ring new_ring, void * map_func_ptr, int * par_perm) {
         nMapFunc map_func = reinterpret_cast<nMapFunc>(map_func_ptr);
         return p_PermPoly(p, perm, old_ring, new_ring, map_func);
     });
+    Singular.method("maFindPerm", [](ring src, jlcxx::ArrayRef<int> perm, ring dst,
+                                     jlcxx::ArrayRef<int> par_perm){
+        int *perm1 = (int *)omAlloc0((rVar(src)+1)*sizeof(int));
+        int *par_perm1 = NULL;
+        if (rPar(src)!=0) par_perm1=(int *)omAlloc0((rPar(src)+1)*sizeof(int));
```
Comment (wbhart) [inline deps/src/rings.cpp]: spacing !=, =, +

File: deps/src/rings.cpp [hunk 4]
```diff
@@ -282,10 +282,23 @@ void singular_define_rings(jlcxx::Module & Singular)
...
+        maFindPerm(src->names,  rVar(src),  rParameter(src),  rPar(src),
```
Comment (wbhart) [inline deps/src/rings.cpp]: single space after comma

File: deps/src/rings.cpp [hunk 5]
```diff
@@ -282,10 +282,23 @@ void singular_define_rings(jlcxx::Module & Singular)
...
+                 dst->names, rVar(dst), rParameter(dst), rPar(dst),
+                 perm1, par_perm1, dst->cf->type);
+        for(int i=0; i<rVar(src); i++)
```
Comment (wbhart) [inline deps/src/rings.cpp]: spacing

File: deps/src/rings.cpp [hunk 6]
```diff
@@ -282,10 +282,23 @@ void singular_define_rings(jlcxx::Module & Singular)
...
+        {perm.push_back(perm1[i]);}
```
Comment (wbhart) [inline deps/src/rings.cpp]: braces on new line

File: deps/src/rings.cpp [hunk 7]
```diff
@@ -282,10 +282,23 @@ void singular_define_rings(jlcxx::Module & Singular)
...
+        for(int j=0; j<rPar(src); j++)
+        {par_perm.push_back(par_perm1[j]);}
```
Comment (wbhart) [inline deps/src/rings.cpp]: braces

File: docs/src/ideal.md [hunk 8]
```diff
@@ -414,3 +422,29 @@ I = Ideal(R, y, x^2, (1 + y^3) * (x^2 - y))
...
+all maximal independent sets for monomial ideals.
+If any ideal $I$ is passed to the function, the computation is performed on
+the leading ideal of $I$.
+
+```@docs
+maximal_independent_set(::sideal; ::Bool)
+```
+
+```julia
+R, (x,y,u,v,w) = PolynomialRing(QQ, ["x", "y", "u", "v", "w"])
```
Comment (wbhart) [inline docs/src/ideal.md]: spaces ,

File: src/ideal/ideal.jl [hunk 9]
```diff
@@ -72,6 +72,21 @@ iszero(I::sideal) = Bool(libSingular.idIs0(I.ptr))
...
 iszerodim(I::sideal) = Bool(libSingular.id_IsZeroDim(I.ptr, base_ring(I).ptr))
 
+@doc Markdown.doc"""
+    dimension(I::sideal{T})
+> Given an ideal $I$ this function computes the Krull dimension
+> of the ring $R/I$, where $R$ is the polynomial ring over
+> which $I$ is an ideal. The ideal must be over a polynomial ring
+> over a field, and a Groebner basis.
+"""
+function dimension(I::sideal{T}) where T <: Singular.RingElem
```
Comment (wbhart) [inline src/ideal/ideal.jl]: why not have one version with T <: Singular.FieldElem which is documented as such and another one for Singular.RingElement which only raises an error (if you really require this)

File: src/ideal/ideal.jl [hunk 10]
```diff
@@ -72,6 +72,21 @@ iszero(I::sideal) = Bool(libSingular.idIs0(I.ptr))
 """
 iszerodim(I::sideal) = Bool(libSingular.id_IsZeroDim(I.ptr, base_ring(I).ptr))
 
+@doc Markdown.doc"""
+    dimension(I::sideal{T})
```
Comment (wbhart) [inline src/ideal/ideal.jl]: prototype should match function

File: src/ideal/ideal.jl [hunk 11]
```diff
@@ -663,4 +678,46 @@ function minimal_generating_set(I::sideal)
...
+function maximal_independent_set(I::sideal{T}; all::Bool = false) where T <: Singular.RingElem
+   I.isGB == false && error("I needs to be a Gröbner basis.")
+   R = base_ring(I)
+   !(typeof(base_ring(R)) <: Singular.Field) && 
+       error("Polynomial ring has to be over a Singular field.")
+   n = nvars(R)
+   a = Array{Int32, 1}()
+   libSingular.scIndIndset(I.ptr, R.ptr, a, all)
+   if all == true
+      m = Int(length(a)/n)
```
Comment (wbhart) [inline src/ideal/ideal.jl]: div, not /

File: src/ideal/ideal.jl [hunk 12]
```diff
@@ -663,4 +678,46 @@ function minimal_generating_set(I::sideal)
...
+      a = Int.(transpose(reshape(a, n, m)))
+      P = Array{Array{spoly, 1}, 1}()
+      for i in 1:m
+         b  = Array{spoly, 1}()
+         for j in findall(x->x==1, a[i, :])
```
Comment (wbhart) [inline src/ideal/ideal.jl]: spacing ==

File: src/ideal/ideal.jl [hunk 13]
```diff
@@ -663,4 +678,46 @@ function minimal_generating_set(I::sideal)
...
+            push!(b, gen(R, j))
+         end
+         push!(P, b)
+      end
+      return P
+   else
+      P = Array{spoly, 1}()
+      for j in findall(x->x==1, a)
```
Comment (wbhart) [inline src/ideal/ideal.jl]: spacing

File: src/ideal/ideal.jl [hunk 14]
```diff
@@ -663,4 +678,46 @@ function minimal_generating_set(I::sideal)
...
+    maximal_independent_set{I::sideal{T}; all::Bool = false)
+> Returns, by default, an array containing a maximal independet set of
+> $lead(I)$. $I$ has to be given by a Gröbner basis.
+> If the additional parameter "all" is set to true, an array containing
+> all maximal independent sets iof $lead(I)$ is returned.
+"""
+function maximal_independent_set(I::sideal{T}; all::Bool = false) where T <: Singular.RingElem
+   I.isGB == false && error("I needs to be a Gröbner basis.")
+   R = base_ring(I)
+   !(typeof(base_ring(R)) <: Singular.Field) &&
```
Comment (wbhart) [inline src/ideal/ideal.jl]: same comments regarding Singular.FieldElem/RingElem

File: src/ideal/ideal.jl [hunk 15]
```diff
@@ -663,4 +678,46 @@ function minimal_generating_set(I::sideal)
...
 end
+
+###############################################################################
+#
+#   Maximal independent set
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    maximal_independent_set{I::sideal{T}; all::Bool = false)
```
Comment (wbhart) [inline src/ideal/ideal.jl]: prototype should match

File: test/ideal/sideal-test.jl [hunk 16]
```diff
@@ -318,3 +320,16 @@ end
    @test f == y^4
 end
 
+@testset "sideal.independentset..." begin
```
Comment (wbhart) [inline test/ideal/sideal-test.jl]: underscore

File: test/ideal/sideal-test.jl [hunk 17]
```diff
@@ -318,3 +320,16 @@ end
...
+   R, (x, y, u, v, w) = PolynomialRing(QQ, ["x", "y", "u", "v", "w"])
+
+   I = Ideal(R, x*y*w, y*v*w, u*y*w, x*v)
+
+   L1 = maximal_independent_set(std(I))
+
+   L2 = maximal_independent_set(std(I), all = true)
+
+   @test L1 == [x, y, u]
+   @test L2 == [[x,y,u], [y,u,v], [x,u,w], [u,v,w]]
```
Comment (wbhart) [inline test/ideal/sideal-test.jl]: spacing ,

**Comment by wbhart** [comment]
do you know what is causing the test failures?

File: src/ideal/ideal.jl [hunk 18]
```diff
@@ -663,4 +676,64 @@ function minimal_generating_set(I::sideal)
...
+end
+
+@doc Markdown.doc"""
+    maximal_independent_set(I::sideal{S}; all::Bool = false) where S <: Union{spoly{T}, spoly{n_unknown{T}}} where T <: Union{Singular.FieldElem, Nemo.FieldElem}
+> Returns, by default, an array containing a maximal independet set of
+> $lead(I)$. $I$ has to be given by a Gröbner basis.
+> If the additional parameter "all" is set to true, an array containing
+> all maximal independent sets iof $lead(I)$ is returned.
+"""
+function maximal_independent_set(I::sideal{S}; all::Bool = false) where S <: Union{spoly{T}, spoly{n_unknown{T}}} where T <: Union{Singular.FieldElem, Nemo.FieldElem}
```
Comment (wbhart) [inline src/ideal/ideal.jl]: I think you want S <: Union{spoly{T}, spoly{n_unknown{U}} where {T <: Singular.FieldElem, U <: Nemo.FieldElem}

File: src/ideal/ideal.jl [hunk 19]
```diff
@@ -663,4 +676,64 @@ function minimal_generating_set(I::sideal)
...
- 
+
+###############################################################################
+#
+#   Independent sets
+#
+###############################################################################
+
+@doc Markdown.doc"""
+    independent_sets(I::sideal{S}) where S <: Union{spoly{T}, spoly{n_unknown{T}}} where T <: Union{Singular.FieldElem, Nemo.FieldElem}
```
Comment (wbhart) [inline src/ideal/ideal.jl]: as per below

File: src/ideal/ideal.jl (hunk 18 above)
Comment (raulepure) [reply src/ideal/ideal.jl]: That's true. I will fix it and prepare the next pull request.

**Pull request merged**